url9d32402837_dusan_rychnovsky_chess_endgames_tgz-pJ8-cz_dusanrychnovsky_chessendgames_core_strategies_PrecomputedValuesJ8/src/cz/dusanrychnovsky/chessendgames/CLIAdapter.java

```diff
@@ -0,0 +1,267 @@
+	 * @param out
+	 */
+	public CLIAdapter(Strategy strategy, BufferedReader in, BufferedWriter out) {
+		this.strategy = strategy;
+		this.in = in;
+		this.out = out;
+	}
+	
+	/**
+	 * 
+	 * @throws IOException
+	 */
+	public void run() throws IOException
+	{
+		printIntroduction();
+		
+		Situation situation = setUpInitialSituation();
+		printLine("Current situation: " + situation);
+		
+		Game game = new Game(strategy, situation, blackPlayer);
+		
+		while (!situation.isFinal())
+		{
+			Move move = getMove(situation);
+			
+			situation = game.doMove(move);
+			if (situation.isFinal()) {
+				break;
+			}
+			
+			situation = game.playMove();
+			printLine("Current situation: " + situation);
+		}
+		
+		Result result = situation.getResult();
+		printLine("Result: " + result);
+	}
+	
+	/**
+	 * 
+	 * @param situation
+	 * @return
+	 * @throws IOException
+	 */
+	private Move getMove(Situation situation) throws IOException 
+	{
+		printLine("Please give your next move ([from] - [to]):");
+		
+		while (true)
+		{
+			String line = in.readLine();
+			
+			try 
+			{
+				String[] tokens = line.split("-");
+				if (tokens.length != 2) {
+					throw new IllegalArgumentException(
+						"Invalid move definition [" + line + "]."
+					);
+				}
+				
+				Position from = Position.get(tokens[0]);
+				Piece piece = situation.getPiece(from);
+				
+				Position to = Position.get(tokens[1]);
+				
+				return new Move(piece, from, to);
+			}
+			catch (IllegalArgumentException ex) {
+				printLine("The given move is invalid. Try again, please.");
+			}
+		}
+	}
+	
+	/**
+	 * 
+	 * @return
+	 * @throws IOException
+	 */
+	private Situation setUpInitialSituation() throws IOException
+	{
+		Situation initialSituation = new Situation();
+		
+		setUpWhiteKing(initialSituation);
+		setUpBlackKing(initialSituation);
+		setUpBlackRook(initialSituation);
+		
+		return initialSituation;
+	}
+	
+	/**
+	 * 
+	 * @param situation
+	 * @throws IOException
+	 */
+	private void setUpBlackRook(Situation situation) throws IOException
+	{
+		printLine("Please give black rook coordinates:");
+		
+		Position position = readPosition();		
+		situation.addPiece(blackRook, position);
+	}
+	
+	/**
+	 * 
+	 * @param situation
+	 * @throws IOException
+	 */
+	private void setUpBlackKing(Situation situation) throws IOException 
+	{
+		printLine("Please give black king coordinates:");
+		
+		Position position = readPosition();		
+		situation.addPiece(blackKing, position);
+	}
+	
+	/**
+	 * 
+	 * @param situation
+	 * @throws IOException
+	 */
+	private void setUpWhiteKing(Situation situation) throws IOException 
+	{
+		printLine("Please give white king coordinates:");
+		
+		Position position = readPosition();		
+		situation.addPiece(whiteKing, position);
+	}
+	
+	/**
+	 * 
+	 * @return
+	 * @throws IOException
+	 */
+	private Position readPosition() throws IOException
+	{
+		while (true)
+		{
+			String line = in.readLine();
+			
+			try {
+				Position result = Position.get(line);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:** 
The loop containing this operation may iterate indefinitely if the operation throws an exception on multiple instances of its evaluation. As the operation throws an exception, it bypasses exit points in the loop and continues to the next iteration through a catch clause.

**Fix:** 
Add a retry counter to ensure that the loop terminates after a predefined number of iterations.